sunbeam-python/sunbeam/core/watcher.py

```diff
@@ -0,0 +1,175 @@
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME,
+        description="Audit template for workload balancing",
+        goal=WORKLOAD_BALANCING_GOAL_NAME,
+        strategy=WORKLOAD_BALANCING_STRATEGY_NAME,
+    )
+    return template
+
+
+def get_enable_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_host_maintenance_audit_template(client=client)
+    return template
+
+
+def get_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_workload_balancing_audit_template(client=client)
+    return template
+
+
+@timeout_decorator.timeout(TIMEOUT)
+def create_audit(
+    client: watcher_client.Client,
+    template: watcher.AuditTemplate,
+    audit_type: str = "ONESHOT",
+    parameters: dict[str, Any] = {},
+) -> watcher.Audit:
+    audit = client.audit.create(
+        audit_template_uuid=template.uuid,
+        audit_type=audit_type,
+        parameters=parameters,
+    )
+    while True:
+        audit_details = client.audit.get(audit.uuid)
+        if audit_details.state in ["SUCCEEDED", "FAILED"]:
+            break
+        time.sleep(SLEEP_INTERVAL)
+    if audit_details.state == "SUCCEEDED":
+        LOG.debug(f"Create Watcher audit {audit.uuid} successfully")
+    else:
+        LOG.debug(f"Create Watcher audit {audit.uuid} failed")
+        raise SunbeamException(
+            f"Create watcher audit failed, template: {template.name}"
+        )
+
+    _check_audit_plans_recommended(client=client, audit=audit)
+    return audit
+
+
+def _check_audit_plans_recommended(client: watcher_client.Client, audit: watcher.Audit):
+    action_plans = client.action_plan.list(audit=audit.uuid)
+    # Verify all the action_plan's state is RECOMMENDED
+    if not all(plan.state in ["RECOMMENDED", "SUCCEEDED"] for plan in action_plans):
+        raise SunbeamException(
+            f"Not all action plan for audit({audit.uuid}) is RECOMMENDED"
+        )
+
+
+def get_actions(
+    client: watcher_client.Client, audit: watcher.Audit
+) -> list[watcher.Action]:
+    """Get list of actions by audit."""
+    return client.action.list(audit=audit.uuid, detail=True)
+
+
+def exec_audit(client: watcher_client.Client, audit: watcher.Audit):
+    """Run audit's action plans."""
+    action_plans = client.action_plan.list(audit=audit.uuid)
+    for action_plan in action_plans:
+        _exec_plan(client=client, action_plan=action_plan)
+    LOG.info(f"All Action plan for Audit {audit.uuid} execution successfully")
+
+
+@timeout_decorator.timeout(TIMEOUT)
```

Comment:
Each actionplan corresponds to multiple actions and each action can be migration. If there are more VMs to be migrated, then the actionplan may take more than TIMEOUT i.e., 3 minutes.

Reply:
The expected behavior is as follows:

- If the migration times out, the command should fail.
- When the user re-runs the command, it should fail if any VMs are still migrating, displaying a message indicating that some instances are in the migration process.
- Once all migration watcher actions are complete, re-running the command should result in success.